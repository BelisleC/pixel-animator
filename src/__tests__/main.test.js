import PixelAnimator from '../main';

it('should iterate over frames', () => {
  const frameIterator = PixelAnimator({
    columns: 2,
    rows: 2,
    frames: 2,
    colorize: ({ column, row, frame, columns, rows, frames }) => ({
      red: column / columns,
      green: row / rows,
      blue: frame / frames,
    }),
  });
  const frame1 = frameIterator();
  const frame2 = frameIterator();
  const frame3 = frameIterator();
  expect([frame1, frame2]).toMatchSnapshot();
  expect(frame1).toEqual(frame3);
});

it('should maintain state between frames', () => {
  const frameIterator = PixelAnimator({
    columns: 2,
    rows: 2,
    frames: 2,
    evolve: ({ column, row, frame, cells }) => {
      if (frame === 0) return false;
      const self = cells(column, row);
      return !self;
    },
    colorize: ({ cell }) => (cell ? 1 : 0),
  });
  const frame1 = frameIterator();
  const frame2 = frameIterator();
  expect([frame1, frame2]).toMatchSnapshot();
});

it('should wrap when referencing out of bounds cell', () => {
  const frameIterator = PixelAnimator({
    columns: 2,
    rows: 2,
    evolve: ({ column, row, frame, cells }) => {
      if (frame === 0) return [column, row];
      return cells(column - 1, row + 1);
    },
    colorize: ({ cell }) => {
      return { red: cell[0] / 2, green: cell[1] / 2 };
    },
  });
  const frame1 = frameIterator();
  const frame2 = frameIterator();
  expect([frame1, frame2]).toMatchSnapshot();
});

it('should normalize color values', () => {
  expect(
    PixelAnimator({
      columns: 1,
      rows: 1,
      colorize: () => 'red',
    })(),
  ).toMatchSnapshot('using a color name');

  expect(
    PixelAnimator({
      columns: 1,
      rows: 1,
      colorize: () => '#FF0000',
    })(),
  ).toMatchSnapshot('using a hex code');

  expect(
    PixelAnimator({
      columns: 1,
      rows: 1,
      colorize: () => ({ red: 1 }),
    })(),
  ).toMatchSnapshot('using a partial RGBA object');

  expect(
    PixelAnimator({
      columns: 1,
      rows: 1,
      colorize: () => ({ red: 0.1, green: 0.2, blue: 0.3, alpha: 0.4 }),
    })(),
  ).toMatchSnapshot('using a complete RGBA object');

  expect(
    PixelAnimator({
      columns: 1,
      rows: 1,
      colorize: () => 0.5,
    })(),
  ).toMatchSnapshot('using a number');

  expect(
    PixelAnimator({
      columns: 1,
      rows: 1,
      colorize: () => null,
    })(),
  ).toMatchSnapshot('using null');
});

it('should generate HTML when given a DOM node', () => {
  document.body.innerHTML = `<div id="root"></div>`;
  const rootElement = document.getElementById('root');
  const transport = PixelAnimator(
    {
      columns: 2,
      rows: 2,
      frames: 2,
      colorize: ({ column, row, frame, columns, rows, frames }) => ({
        red: column / columns,
        green: row / rows,
        blue: frame / frames,
      }),
    },
    rootElement,
  );
  transport.pause();
  transport.next();
  expect(rootElement).toMatchSnapshot('first frame');
  transport.next();
  expect(rootElement).toMatchSnapshot('second frame');
});
