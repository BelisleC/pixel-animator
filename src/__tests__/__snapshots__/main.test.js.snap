// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should generate HTML when given a DOM node: first frame 1`] = `
<div
  class=" pixel-animator"
  id="root"
  style="width: 320px; height: 320px;"
>
  <div
    class="pa-cell pa-cell-0"
    style="width: 50%; height: 50%; display: inline-block; vertical-align: top; background-color: rgb(0, 0, 0);"
  />
  <div
    class="pa-cell pa-cell-1"
    style="width: 50%; height: 50%; display: inline-block; vertical-align: top; background-color: rgb(127, 0, 0);"
  />
  <div
    class="pa-cell pa-cell-2"
    style="width: 50%; height: 50%; display: inline-block; vertical-align: top; background-color: rgb(0, 127, 0);"
  />
  <div
    class="pa-cell pa-cell-3"
    style="width: 50%; height: 50%; display: inline-block; vertical-align: top; background-color: rgb(127, 127, 0);"
  />
</div>
`;

exports[`should generate HTML when given a DOM node: second frame 1`] = `
<div
  class=" pixel-animator"
  id="root"
  style="width: 320px; height: 320px;"
>
  <div
    class="pa-cell pa-cell-0"
    style="width: 50%; height: 50%; display: inline-block; vertical-align: top; background-color: rgb(0, 0, 127);"
  />
  <div
    class="pa-cell pa-cell-1"
    style="width: 50%; height: 50%; display: inline-block; vertical-align: top; background-color: rgb(127, 0, 127);"
  />
  <div
    class="pa-cell pa-cell-2"
    style="width: 50%; height: 50%; display: inline-block; vertical-align: top; background-color: rgb(0, 127, 127);"
  />
  <div
    class="pa-cell pa-cell-3"
    style="width: 50%; height: 50%; display: inline-block; vertical-align: top; background-color: rgb(127, 127, 127);"
  />
</div>
`;

exports[`should iterate over frames 1`] = `
Array [
  Array [
    Object {
      "alpha": 1,
      "blue": 0,
      "green": 0,
      "red": 0,
    },
    Object {
      "alpha": 1,
      "blue": 0,
      "green": 0,
      "red": 127,
    },
    Object {
      "alpha": 1,
      "blue": 0,
      "green": 127,
      "red": 0,
    },
    Object {
      "alpha": 1,
      "blue": 0,
      "green": 127,
      "red": 127,
    },
  ],
  Array [
    Object {
      "alpha": 1,
      "blue": 127,
      "green": 0,
      "red": 0,
    },
    Object {
      "alpha": 1,
      "blue": 127,
      "green": 0,
      "red": 127,
    },
    Object {
      "alpha": 1,
      "blue": 127,
      "green": 127,
      "red": 0,
    },
    Object {
      "alpha": 1,
      "blue": 127,
      "green": 127,
      "red": 127,
    },
  ],
]
`;

exports[`should maintain state between frames 1`] = `
Array [
  Array [
    Object {
      "alpha": 1,
      "blue": 0,
      "green": 0,
      "red": 0,
    },
    Object {
      "alpha": 1,
      "blue": 0,
      "green": 0,
      "red": 0,
    },
    Object {
      "alpha": 1,
      "blue": 0,
      "green": 0,
      "red": 0,
    },
    Object {
      "alpha": 1,
      "blue": 0,
      "green": 0,
      "red": 0,
    },
  ],
  Array [
    Object {
      "alpha": 1,
      "blue": 255,
      "green": 255,
      "red": 255,
    },
    Object {
      "alpha": 1,
      "blue": 255,
      "green": 255,
      "red": 255,
    },
    Object {
      "alpha": 1,
      "blue": 255,
      "green": 255,
      "red": 255,
    },
    Object {
      "alpha": 1,
      "blue": 255,
      "green": 255,
      "red": 255,
    },
  ],
]
`;

exports[`should wrap when referencing out of bounds cell 1`] = `
Array [
  Array [
    Object {
      "alpha": 1,
      "blue": 0,
      "green": 0,
      "red": 0,
    },
    Object {
      "alpha": 1,
      "blue": 0,
      "green": 0,
      "red": 127,
    },
    Object {
      "alpha": 1,
      "blue": 0,
      "green": 127,
      "red": 0,
    },
    Object {
      "alpha": 1,
      "blue": 0,
      "green": 127,
      "red": 127,
    },
  ],
  Array [
    Object {
      "alpha": 1,
      "blue": 0,
      "green": 127,
      "red": 127,
    },
    Object {
      "alpha": 1,
      "blue": 0,
      "green": 127,
      "red": 0,
    },
    Object {
      "alpha": 1,
      "blue": 0,
      "green": 0,
      "red": 127,
    },
    Object {
      "alpha": 1,
      "blue": 0,
      "green": 0,
      "red": 0,
    },
  ],
]
`;
